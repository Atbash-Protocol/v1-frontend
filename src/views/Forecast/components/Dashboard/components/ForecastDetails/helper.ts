import { EPOCH_PER_DAY, LAMBO_PRICE } from 'views/Forecast/config';

export const computeDailyROI = (rewardYieldPercent: string, duration: number) => {
    return (1 + Number(rewardYieldPercent) / 100) ** (duration * EPOCH_PER_DAY) - 1;
};

export const getBashRewardsEstimation = (stakedSBAmount: string, dailyROI: number) => {
    return Number(stakedSBAmount) * dailyROI;
};

export const getPotentialReturn = (stakedSBAmount: string, bashRewardsEstimation: number, futureBASHMarketPrice: string) =>
    (Number(stakedSBAmount) + bashRewardsEstimation) * Number(futureBASHMarketPrice);

export const getLamboEstimation = (potentialReturn: number) => Math.round(potentialReturn / LAMBO_PRICE);
