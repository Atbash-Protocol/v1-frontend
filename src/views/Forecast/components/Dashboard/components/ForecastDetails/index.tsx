import { Box, Typography } from '@mui/material';
import Decimal from 'decimal.js';
import { useTranslation } from 'react-i18next';

import { theme } from 'constants/theme';
import { formatUSD } from 'helpers/price-units';

import { computeDailyROI, getBashRewardsEstimation, getLamboEstimation, getPotentialReturn } from './helper';

interface ForecastDetailsProps {
    data: Record<string, string>;
    initialPrice: Decimal;
    duration: number;
}

const ForecastDetails = ({ data, duration, initialPrice }: ForecastDetailsProps) => {
    const { t } = useTranslation();
    const { stakedSBAmount, BASHPriceAtPurchase, rewardYieldPercent, futureBASHMarketPrice } = data;

    const dailyROI = computeDailyROI(rewardYieldPercent, duration);
    const bashRewardsEstimation = getBashRewardsEstimation(stakedSBAmount, dailyROI);
    const potentialReturn = getPotentialReturn(stakedSBAmount, bashRewardsEstimation, futureBASHMarketPrice);
    const nbLambo = getLamboEstimation(potentialReturn);

    const computedForecast = [
        { key: 'globe:YourInitialInvestment', value: formatUSD(Number(stakedSBAmount) * Number(BASHPriceAtPurchase), 2) },
        { key: 'globe:CurrentWealth', value: formatUSD(initialPrice.toNumber() * Number(stakedSBAmount), 2) },
        { key: 'globe:BASHRewardsEstimation', value: Intl.NumberFormat('en', { notation: 'compact', maximumFractionDigits: 2 }).format(bashRewardsEstimation) },
        { key: 'globe:PotentialReturn', value: formatUSD(potentialReturn) },
    ].map(({ key, value }, index) => {
        return (
            <Box key={index} sx={{ justifyContent: 'space-between', color: 'white', display: 'inline-flex', width: '100%', mt: theme.spacing(2) }}>
                <Typography variant="body1">{t(key)}</Typography>
                <Typography variant="body1" noWrap>
                    {value}
                </Typography>
            </Box>
        );
    });

    const lambos = (
        <Box key="lambos" sx={{ justifyContent: 'space-between', color: 'white', display: 'inline-flex', width: '100%', mt: theme.spacing(2) }}>
            <Typography variant="body1">{t('globe:PotentialNumberLambo')}</Typography>
            <Typography variant="body1">{nbLambo}</Typography>
        </Box>
    );

    return <Box sx={{ height: '100%', display: 'flex', flexDirection: 'column', alignItems: 'center', justifyContent: 'center' }}>{[...computedForecast, lambos]}</Box>;
};

export default ForecastDetails;
