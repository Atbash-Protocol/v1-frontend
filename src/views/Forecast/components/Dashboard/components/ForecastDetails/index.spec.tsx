import { render } from '@testing-library/react';
import Decimal from 'decimal.js';

import ForecastDetails from '.';
import * as ForecastDetailsHelperModule from './helper';

describe('ForecastDetails', () => {
    it('renders', () => {
        jest.spyOn(ForecastDetailsHelperModule, 'computeDailyROI').mockReturnValue(100);
        jest.spyOn(ForecastDetailsHelperModule, 'getBashRewardsEstimation').mockReturnValue(12000);
        jest.spyOn(ForecastDetailsHelperModule, 'getPotentialReturn').mockReturnValue(12039);
        jest.spyOn(ForecastDetailsHelperModule, 'getLamboEstimation').mockReturnValue(1);

        const props = {
            data: {
                stakedSBAmount: '122.01',
                BASHPriceAtPurchase: '80',
                rewardYieldPercent: '120',
                futureBASHMarketPrice: '1200',
            },
            duration: 10,
            initialPrice: new Decimal(80),
        };

        const container = render(<ForecastDetails {...props} />);

        expect(container).toMatchSnapshot();
    });
});
