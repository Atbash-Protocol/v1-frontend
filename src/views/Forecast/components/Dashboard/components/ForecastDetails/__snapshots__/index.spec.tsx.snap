// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastDetails renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-16rwds4"
      >
        <div
          class="MuiBox-root css-faoohn"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            globe:YourInitialInvestment
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-fvtpii-MuiTypography-root"
          >
            $9,760.80
          </p>
        </div>
        <div
          class="MuiBox-root css-faoohn"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            globe:CurrentWealth
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-fvtpii-MuiTypography-root"
          >
            $9,760.80
          </p>
        </div>
        <div
          class="MuiBox-root css-faoohn"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            globe:BASHRewardsEstimation
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-fvtpii-MuiTypography-root"
          >
            12K
          </p>
        </div>
        <div
          class="MuiBox-root css-faoohn"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            globe:PotentialReturn
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-fvtpii-MuiTypography-root"
          >
            $12,039
          </p>
        </div>
        <div
          class="MuiBox-root css-faoohn"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            globe:PotentialNumberLambo
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            1
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-16rwds4"
    >
      <div
        class="MuiBox-root css-faoohn"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          globe:YourInitialInvestment
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-fvtpii-MuiTypography-root"
        >
          $9,760.80
        </p>
      </div>
      <div
        class="MuiBox-root css-faoohn"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          globe:CurrentWealth
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-fvtpii-MuiTypography-root"
        >
          $9,760.80
        </p>
      </div>
      <div
        class="MuiBox-root css-faoohn"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          globe:BASHRewardsEstimation
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-fvtpii-MuiTypography-root"
        >
          12K
        </p>
      </div>
      <div
        class="MuiBox-root css-faoohn"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          globe:PotentialReturn
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-fvtpii-MuiTypography-root"
        >
          $12,039
        </p>
      </div>
      <div
        class="MuiBox-root css-faoohn"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          globe:PotentialNumberLambo
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          1
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
